### PR TITLE
Fix reference count handling in TransportSingleShardAction

### PR DESCRIPTION
**Pull request by original-brownbear**
Just like in the search service, we need to only release the reference after passing the supplied object to the listener.
Since this is the same pattern in the `TransportSingleShardAction` and in `SearchService` and we will likely have more spots like this in the near future I extracted the logic to a new utility.

> non-issue since this had no impact on released versions yet.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/ActionRunnable.java [hunk 1]
```diff
@@ -58,6 +59,35 @@ public String toString() {
...
     }
 
+    /**
+     * Same as {@link #supply(ActionListener, CheckedSupplier)} but the supplier always returns an object of reference counted result type
+     * which will have its reference count decremented after invoking the listener.
+     * @param listener Listener to invoke
+     * @param supplier Supplier that provides the reference counted value to pass to the listener
+     * @return Wrapped {@code Runnable}
+     */
+    public static <T extends RefCounted> ActionRunnable<T> supplyRefCounted(
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/ActionRunnable.java:69]: Naming nit/suggestion, how about this?

```suggestion
    public static <T extends RefCounted> ActionRunnable<T> supplyAndDecRef(
```

File: server/src/main/java/org/elasticsearch/action/ActionRunnable.java [hunk 2]
```diff
@@ -58,6 +59,35 @@ public String toString() {
         });
     }
 
+    /**
+     * Same as {@link #supply(ActionListener, CheckedSupplier)} but the supplier always returns an object of reference counted result type
+     * which will have its reference count decremented after invoking the listener.
+     * @param listener Listener to invoke
+     * @param supplier Supplier that provides the reference counted value to pass to the listener
+     * @return Wrapped {@code Runnable}
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/ActionRunnable.java:67]: Nit: tbh I find these bits redundant and IMO actually makes the overall doc harder to parse. Suggest dropping them.

```suggestion
```

**Review by DaveCTurner — APPROVED** [review]
LGTM with a couple of superficial suggestions

**Comment by original-brownbear** [comment]
Thanks David!

**Pull request merged**